.github/workflows/build.yml

```diff
@@ -0,0 +1,31 @@
+name: build
+
+on: [push]
+
+env:
+  TAG_GHCR: mrc-ide/wodin
+
+jobs:
+  build-and-push:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v4
+      - name: Setup env and login to GHCR
+        id: ci-env
+        uses: ./.github/actions/setup-and-login
+        with:
+          ghcr-username: ${{ github.actor }}
+          ghcr-password: ${{ secrets.GITHUB_TOKEN }}
+      
+      - name: Build and push docker
+        uses: docker/build-push-action@v5
+        with:
+          file: ./docker/Dockerfile
+          push: true
+          tags: |
+            ghcr.io/${{env.TAG_GHCR}}:${{steps.ci-env.outputs.CI_SHA}}
+            ghcr.io/${{env.TAG_GHCR}}:${{steps.ci-env.outputs.CI_BRANCH}}
```

Comment:
So this workflow always runs, and always pushes images of both sha and branch tags regardless of test status. As opposed to previous way where no images were pushed if any e2e tests failed.. 
I wonder if we might want to be a bit more cautious, in main branch particularly, just in case we accidentally break things on a merge. 
Could have a "latest" tag which only gets pushed if all tests in all other jobs have passed maybe?

Reply:
this one took a little bit of fiddling around but have done it! the strategy i went for was to get the manifest for the docker image and upload the same manifest to another tag, this way we dont need to rebuild the entire image after the testing is done

this job will only run on `main` and only after all other checks are finished, ive had to combine the build workflow into the same file as the testing workflow because we need to conditionally wait for jobs to finish and having them as different jobs in the same workflow file makes this easy